File: leetcode/python/0020. Valid Parentheses.py
# Time complexity: O(N)
# Space complexity: O(N)

class Solution(object):
    def isValid(self, s):
        opening = ['(', '{', '[']

        stack = []

        for bracket in s:
            if bracket in opening:
                stack.append(bracket)
            else:
                if len(stack) == 0:
                    return False
                brack = stack.pop()
                if bracket == ')' and brack != '(':
                    return False
                elif bracket == '}' and brack != '{':
                    return False
                elif bracket == ']' and brack != '[':
                    return False
        
        return len(stack) == 0

File: leetcode/java/0020. Valid Parentheses.java
package leetcode.java;

import java.util.Stack;

// Time complexity: O(N)
// Space complexity: O(N)

class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack =  new Stack<>();
        for (Character c : s.toCharArray()) {
            if (c == '(' || c == '{' || c == '[') {
                stack.add(c);
            } else {
                if (stack.size() == 0) {
                    return false;
                }
                char temp = stack.pop();
                if (temp == '(' && c != ')') {
                    return false;
                } else if (temp == '{' && c != '}') {
                    return false;
                } else if (temp == '[' && c != ']') {
                    return false;
                }
            }
        }
        return stack.size() == 0;
    }
}